newton/examples/example_robot_manipulating_cloth.py

```diff
@@ -0,0 +1,571 @@
+
+        # 4. Define a null-space velocity term pulling joints toward q_des
+        #    K_null is a small gain so it doesn't override main task
+        K_null = 1.0
+        delta_q_null = K_null * (q_des - q)
+
+        # 5. Combine primary task and null-space controller
+        delta_q = J_inv @ delta_target + N @ delta_q_null
+
+        # Apply gripper finger control
+        delta_q[-2] = self.target[-1] * 4 - q[-2]
+        delta_q[-1] = self.target[-1] * 4 - q[-1]
+
+        return delta_q
+
+    def run(self):
+        for frame_idx in range(self.num_frames):
+            self.advance_frame()
+
+            if self.add_cloth and not (frame_idx % 10):
+                self.cloth_solver.rebuild_bvh(self.state_0)
+                self.capture_cuda_graph()
+
+            if self.renderer is not None:
+                self.renderer.begin_frame(self.sim_time)
+                self.renderer.render(self.state_0)
+                self.renderer.end_frame()
+
+    def cloth_sim_substep(self, state_in, state_out):
+        self.cloth_solver.step(self.model, state_in, state_out, self.control, None, self.sim_dt)
+
+    def advance_frame(self):
+        target_joint_qd = self.get_target_joint_q(self.state_0)
+        for step in range(self.num_substeps):
+            # robot sim
+            self.state_0.clear_forces()
+            self.state_1.clear_forces()
+
+            if self.add_robot:
+                particle_count = self.model.particle_count
+                # set particle_count = 0 to circumvent
+                self.model.particle_count = 0
+                self.model.gravity = wp.vec3(0)
+
+                # Update the robot pose - this will modify state_0 and copy to state_1
+                self.model.shape_contact_pair_count = 0
+
+                self.state_0.joint_qd.assign(target_joint_qd)
+                # Just update the forward kinematics to get body positions from joint coordinates
+                self.robot_solver.step(self.model, self.state_0, self.state_1, self.control, None, self.sim_dt)
+
+                self.state_0.particle_f.zero_()
+
+                self.model.particle_count = particle_count
+                self.model.gravity = wp.vec3(0, -1000, 0)
+
+            # # cloth sim
+            collide(self.model, self.state_0)
+
+            if self.add_cloth:
+                if self.use_cuda_graph:
+                    if step % 2:
+                        # odd step, state_1 in state_0 out
+                        wp.capture_launch(self.cuda_graph_odd_step)
+                    else:
+                        # even step, state_0 in state_1 out
+                        wp.capture_launch(self.cuda_graph_even_step)
+                else:
+                    self.cloth_sim_substep(self.state_0, self.state_1)
+
+            self.state_0, self.state_1 = self.state_1, self.state_0
+
+            self.sim_time += self.sim_dt
+
+
+if __name__ == "__main__":
+    with wp.ScopedDevice(device=wp.get_device("cuda")):
```

Comment:
I think we don't need a `ScopedDevice` here to fix it to CUDA? Warp uses CUDA by default, you could just remove this.